src/Microsoft.AspNet.OData.Shared/Query/DefaultSkipTokenHandler.cs

```diff
@@ -0,0 +1,387 @@
+            return ExpressionHelpers.Where(query, whereLambda, query.ElementType);
+        }
+
+        /// <summary>
+        /// Generates a dictionary with property name and property values specified in the skiptoken value.
+        /// </summary>
+        /// <param name="value">The skiptoken string value.</param>
+        /// <param name="context">The <see cref="ODataQueryContext"/> which contains the <see cref="IEdmModel"/> and some type information</param>
+        /// <returns>Dictionary with property name and property value in the skiptoken value.</returns>
+        protected static IDictionary<string, object> PopulatePropertyValuePairs(string value, ODataQueryContext context)
+        {
+            Contract.Assert(context != null);
+
+            IDictionary<string, object> propertyValuePairs = new Dictionary<string, object>();
+            string[] keyValues = value.Split(CommaDelimiter);
+            foreach (string keyAndValue in keyValues)
+            {
+                string[] pieces = keyAndValue.Split(new char[] { propertyDelimiter }, 2);
+                if (pieces.Length > 1 && !String.IsNullOrWhiteSpace(pieces[1]))
+                {
+                    object propValue = null;
+                    if (pieces[1].StartsWith("'enumType'"))
+                    {
+                        string enumValue = pieces[1].Remove(0, 10);
+                        IEdmTypeReference type = EdmLibHelpers.GetTypeReferenceOfProperty(context.Model, context.ElementClrType, pieces[0]);
+                        propValue = ODataUriUtils.ConvertFromUriLiteral(enumValue, ODataVersion.V401, context.Model, type);
+                    }
+                    else
+                    {
+                        propValue = ODataUriUtils.ConvertFromUriLiteral(pieces[1], ODataVersion.V401);
+                    }
+                    if (!String.IsNullOrWhiteSpace(pieces[0]))
+                    {
+                        propertyValuePairs.Add(pieces[0], propValue);
+                    }
+                }
+            }
+
+            return propertyValuePairs;
+        }
+
+        /// <summary>
+        /// Returns the list of properties that should be used for generating the skiptoken value. 
+        /// </summary>
+        /// <param name="lastMember">The last record that will be returned in the response.</param>
+        /// <param name="model">IEdmModel</param>
+        /// <param name="orderByNodes">OrderBy nodes in the original request.</param>
+        /// <returns>List of properties that should be used for generating the skiptoken value.</returns>
+        protected static IEnumerable<IEdmProperty> GetPropertiesForSkipToken(object lastMember, IEdmModel model, IList<OrderByNode> orderByNodes)
+        {
+            IEdmType edmType = GetTypeFromObject(lastMember, model);
+            IEdmEntityType entity = edmType as IEdmEntityType;
+            if (entity == null)
+            {
+                return null;
```

Comment:
>return null; [](start = 16, length = 12)

throw exception? or return an empty?

Reply:
We want to return an empty list not an exception but why instantiate a new List? 

---
In reply to: [282573265](https://github.com/OData/WebApi/pull/1724#discussion_r282573265) [](ancestors = 282573265)